news/_posts/2019-10-10-about-overriding-system.md

```diff
@@ -0,0 +1,131 @@
+---
+layout: post
+title:  "the future of the overriding system of PrestaShop"
+subtitle: "From 2010 to 2019, what we have learned"
+date:   2019-10-10 10:30:00
+authors: [ mickaelandrieu ]
+icon: icon-calendar
+tags:
+ - core
+ - hooks
+---
+
+# About the future of the overriding system of PrestaShop
+
+TLDR; the system is deprecated and not usable anymore in the back office and deprecated/discouraged in the front office. It will be removed in the next major version of PrestaShop in its original form.
+
+
+## What are the overrides?
+
+Introduced in PrestaShop 1.4.x almost 9 years ago (in 2010 for the first alphas!), the overrides allows developers to replace PrestaShop Core classes by their own in the Shop and since 1.6.x in the modules.
+
+At the time, the idea was to give a better solution to developers that used to modify directy the sources of PrestaShop to alter the right behavior for their needs.
+With overrides, the modified files were located in a specific folder, both at the Shop or Module level. This way, upgrading the shops were easier: no need to care about these modified files anymore, "PrestaShop" was able to deal with them on their own.
+
+### Please, show me an exemple!
+
+A [recent request](https://github.com/PrestaShop/PrestaShop/pull/11910) from a community member was to update some classes to allow module developers to alter both the Customer form and Customer Form formatter.
+
+We provide native (or Core) classes in PrestaShop, named *CustomerForm* and *FormFormatter* but our implementation doesn't fits his client needs. So his strategy was to override the 2 classes and use their own, extending the core class to not have to rewrite everything.
+
+We didn't discuss details that much the developers needs, but we can imagine he wants to be able to override this function in charge of defining the Customer form:
+
+```php
+<?php
+// override/classes/form/CustomerFormatter
+use CustomerFormatter as NativeCustomerFormatter;
+
+class CustomerFormatter extends NativeCustomerFormatter
+{
+    public function getFormat()
+    {
+        // Here is the business logic to create the form
+        // with the right fields.
+    }
+}
+```
+
+Doing that allow this class to be used as a replacement of the original class in the Core.
+
+But... the core implementation is not "extensible" enough: this is a bug! Most of its properties are private and a function named `addConstraints` (in charge of  the Form validation) is private!
```

Comment:
Not sure it's a bug, this is something they wanted at first. It's a bad design conception.